perun-registrar-lib/src/main/java/cz/metacentrum/perun/registrar/impl/AppAutoRejectionScheduler.java

```diff
@@ -0,0 +1,246 @@
+		} catch(InternalErrorException | PerunException e){
+			log.error("Synchronizer: voApplicationsAutoRejection", e);
+		}
+
+		List<Group> groups = getAllEligibleGroups();
+		// check applications expiration in eligible groups
+		try {
+			groupApplicationsAutoRejection(groups);
+		} catch (InternalErrorException | PerunException e){
+			log.error("Synchronizer: groupApplicationsAutoRejection", e);
+		}
+	}
+
+	/**
+	 * Returns current system time.
+	 *
+	 * @return current time.
+	 */
+	public LocalDate getCurrentLocalDate() {
+		return LocalDate.now();
+	}
+
+	/**
+	 * Checks all applications for given vos and if some application is expired (according to expiration rules set by
+	 * VO Manager), then rejects this application.
+	 *
+	 * @param vos eligible virtual organizations
+	 * @throws PerunException perun exception
+	 */
+	private void voApplicationsAutoRejection(List<Vo> vos) throws PerunException {
+		List<String> states = new ArrayList<>();
+		states.add("NEW");
+		states.add("VERIFIED");
+
+		for (Vo vo : vos) {
+			Attribute expiration = perun.getAttributesManagerBl().getAttribute(sess, vo, A_VO_APP_EXP_RULES);
+			if (expiration.getValue() != null) {
+				List<Application> applications = registrarManager.getApplicationsForVo(sess, vo, states);
+				rejectExpiredApplications(applications, expiration);
+			}
+		}
+	}
+
+	/**
+	 * Gets all existing groups and then checks all applications for this groups and if some application is expired
+	 * (according to expiration rules set by VO Manager), then rejects this application.
+	 *
+	 * @param groups eligible groups
+	 * @throws PerunException perun exception
+	 */
+	private void groupApplicationsAutoRejection(List<Group> groups) throws PerunException {
+		List<String> states = new ArrayList<>();
+		states.add("NEW");
+		states.add("VERIFIED");
+
+		for (Group group : groups) {
+			Attribute expiration = perun.getAttributesManagerBl().getAttribute(sess, group, A_GROUP_APP_EXP_RULES);
+			if (expiration.getValue() != null) {
+				List<Application> applications = registrarManager.getApplicationsForGroup(sess, group, states);
+				rejectExpiredApplications(applications, expiration);
+			}
+		}
+	}
+
+	/**
+	 * Compares date of last modification of application to values in expiration attribute and if finds expired application, then
+	 * rejects it.
+	 *
+	 * @param applications applications
+	 * @param expiration attribute with number of days to application expiration
+	 */
+	private void rejectExpiredApplications (List<Application> applications, Attribute expiration) {
+		Map<String, String> attrValue = expiration.valueAsMap();
+		for(Application application : applications) {
+			String date = application.getModifiedAt();
+			LocalDate modifiedAt = LocalDate.parse(date.substring(0, 10));
+			LocalDate now = getCurrentLocalDate();
+			if (application.getState() == Application.AppState.NEW) {
+				int expirationAppWaitingForEmail = Integer.parseInt(attrValue.get("emailVerification"));
```

Comment:
I think this could cause nullPointerException when the "emailVerification" key is not set. This can happen in all places where you get a value for a key from this attribute.

Reply:
Good point. Your implementation below should solve this problem and I also think, that it covers all use cases.